src/main/java/com/google/cloud/tools/app/DeployAction.java

```diff
@@ -0,0 +1,86 @@
+/**
+ * Copyright 2016 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.tools.app;
+
+import com.google.cloud.tools.app.ProcessCaller.Tool;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Deploys an app to the server.
+ */
+public class DeployAction extends Action {
+
+  private static Set<Option> acceptedFlags = ImmutableSet.of(
+      Option.BUCKET,
+      Option.DOCKER_BUILD,
+      Option.FORCE,
+      Option.IMAGE_URL,
+      Option.PROMOTE,
+      Option.SERVER,
+      Option.VERSION
+  );
+  private File stagedDirectory;
+
+  /**
+   * Initialises all the necessary properties to deploy an application.
+   *
+   * @param stagedDirectoryLocation directory where staged application is located
+   * @param flags deployment flags
+   */
+  public DeployAction(String stagedDirectoryLocation, Map<Option, String> flags) {
+    super(flags);
+    Preconditions.checkNotNull(stagedDirectoryLocation);
```

Comment:
So I don't think we should use @NotNull.  It's actively discouraged
internally as it's redundant when you use @Nullable.  i.e. the absence of
@Nullabe implies @NotNull.

There is no code enforcing the annotations (IntelliJ does this for IJ code,
but we won't use that for these projects).  It's purely for the benefit of
readers and tools like findbugs.


On Wed, Mar 23, 2016 at 3:19 PM, João Martins <notifications@github.com>
wrote:

> In src/main/java/com/google/cloud/tools/app/DeployAction.java
> <https://github.com/GoogleCloudPlatform/gcp-tools-lib-for-java/pull/1#discussion_r57220798>
> :
>
> > +      Option.IMAGE_URL,
> > +      Option.PROMOTE,
> > +      Option.SERVER,
> > +      Option.VERSION
> > +  );
> > +  private File stagedDirectory;
> > +
> > +  /**
> > +   * Initialises all the necessary properties to deploy an application.
> > +   *
> > +   * @param stagedDirectoryLocation directory where staged application is located
> > +   * @param flags deployment flags
> > +   */
> > +  public DeployAction(String stagedDirectoryLocation, Map<Option, String> flags) {
> > +    super(flags);
> > +    Preconditions.checkNotNull(stagedDirectoryLocation);
>
> Ah nevermind, I was noobing with IntelliJ.
>
> What is the expected effect of the annotation when you pass a null to a
> @Nonnull <https://github.com/Nonnull> parameter? Should it cause any
> exception to be thrown, or is it just informative?
>
> No exceptions are being thrown in unit tests when I'm passing nulls to
> @Nonnull <https://github.com/Nonnull> parameters.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/gcp-tools-lib-for-java/pull/1/files/10d1211864218e1bd38d247da31c02ff9d402984#r57220798>
>